File: Mongo/queries.js
use capta_tickets

print("\n" + "=".repeat(80))
print("🔍 CONSULTAS PRINCIPALES - SISTEMA CAPTA TICKETS")
print("=".repeat(80))

const startDate = new Date("2025-09-01T00:00:00Z")
const endDate = new Date("2025-10-28T00:00:00Z")
const estadoFiltro = "open"
const clasificadorFiltro = "altos_palmas_torre1"

print("\n📅 PARÁMETROS DE PRUEBA:")
print(`   Fecha inicio: ${startDate.toISOString()}`)
print(`   Fecha fin: ${endDate.toISOString()}`)
print(`   Estado filtro: ${estadoFiltro}`)
print(`   Clasificador filtro: ${clasificadorFiltro}`)

print("\n" + "=".repeat(80))
print("CONSULTA 1: LISTA DE CASOS")
print("=".repeat(80))

print("\n📖 SEMÁNTICA:")
print("   Retorna tickets que cumplan TODOS los filtros aplicables:")
print("   - Estuvieron ACTIVOS (no cerrados) durante [startDate, endDate)")
print("   - Tienen el estado especificado al final del período (endDate)")
print("   - Están clasificados en el nodo especificado o sus descendientes")

print("\n🔧 IMPLEMENTACIÓN:")

function getDescendantClassifiers(nodeId) {
    const node = db.classifiers.findOne({ _id: nodeId })
    if (!node) return [nodeId]

    const descendants = db.classifiers.find({
        ancestors: nodeId
    }).toArray().map(c => c._id)

    return [nodeId, ...descendants]
}

function getStateAtEndDate(ticket, endDate) {
    const history = ticket.history || []

    for (let i = history.length - 1; i >= 0; i--) {
        const action = history[i]
        if (action.action === 'state_change' && action.timestamp < endDate) {
            return action.to
        }
    }

    return ticket.history[0]?.details?.initialState || ticket.currentState
}

const descendientes = getDescendantClassifiers(clasificadorFiltro)
print(`\n   Nodos incluidos en filtro jerárquico: ${JSON.stringify(descendientes)}`)

const pipeline1 = [
    {
        $match: {
            createdAt: { $lt: endDate },
            $or: [
                { closedAt: { $exists: false } },
                { closedAt: { $gte: startDate } }
            ],
            "currentClassifications.tipo_solicitud": { $in: descendientes }
        }
    },
    {
        $project: {
            _id: 1,
            title: 1,
            currentState: 1,
            currentClassifications: 1,
            createdAt: 1,
            closedAt: 1,
            history: 1
        }
    }
]

let resultadosQ1 = db.tickets.aggregate(pipeline1).toArray()

resultadosQ1 = resultadosQ1.filter(ticket => {
    const stateAtEnd = getStateAtEndDate(ticket, endDate)
    return stateAtEnd === estadoFiltro
})

print(`\n✅ RESULTADO: ${resultadosQ1.length} tickets encontrados`)

print("\n📋 Primeros 3 tickets:")
resultadosQ1.slice(0, 3).forEach((ticket, idx) => {
    const stateAtEnd = getStateAtEndDate(ticket, endDate)
    print(`\n   ${idx + 1}. ID: ${ticket._id}`)
    print(`      Título: ${ticket.title}`)
    print(`      Estado actual: ${ticket.currentState}`)
    print(`      Estado al ${endDate.toISOString().split('T')[0]}: ${stateAtEnd}`)
    print(`      Clasificación: ${ticket.currentClassifications.tipo_solicitud}`)
    print(`      Creado: ${ticket.createdAt.toISOString().split('T')[0]}`)
})

print("\n" + "=".repeat(80))
print("CONSULTA 2: CANTIDAD DE REAPERTURAS")
print("=".repeat(80))

print("\n📖 SEMÁNTICA:")
print("   Cuenta el TOTAL de reaperturas en el sistema")
print("   Una reapertura = transición de 'closed' a cualquier estado activo")
print("   Un ticket puede tener múltiples reaperturas")

print("\n🔧 IMPLEMENTACIÓN:")
print("   Método 1: Usando campo desnormalizado 'reopenCount'")

const pipeline2a = [
    {
        $group: {
            _id: null,
            totalReaperturas: { $sum: "$reopenCount" }
        }
    }
]

const resultado2a = db.tickets.aggregate(pipeline2a).toArray()
const totalReaperturas = resultado2a[0]?.totalReaperturas || 0

print(`\n✅ RESULTADO (Método 1 - Desnormalizado): ${totalReaperturas} reaperturas`)

print("\n   Método 2: Calculando desde el histórico (para verificación)")

const pipeline2b = [
    { $unwind: "$history" },
    {
        $match: {
            "history.action": "state_change",
            "history.from": "closed",
            "history.to": { $ne: "closed" }
        }
    },
    { $count: "totalReaperturas" }
]

const resultado2b = db.tickets.aggregate(pipeline2b).toArray()
const totalReaperturasVerificado = resultado2b[0]?.totalReaperturas || 0

print(`\n✅ RESULTADO (Método 2 - Histórico): ${totalReaperturasVerificado} reaperturas`)

if (totalReaperturas === totalReaperturasVerificado) {
    print("\n   ✓ Ambos métodos coinciden (datos consistentes)")
} else {
    print("\n   ⚠ Los métodos NO coinciden (revisar consistencia)")
}

const ticketsConReaperturas = db.tickets.countDocuments({ reopenCount: { $gt: 0 } })
print(`\n   📊 Tickets con al menos 1 reapertura: ${ticketsConReaperturas}`)

const maxReaperturas = db.tickets.findOne({}, { sort: { reopenCount: -1 } })
print(`   📊 Máximo de reaperturas en un ticket: ${maxReaperturas?.reopenCount || 0}`)

print("\n" + "=".repeat(80))
print("CONSULTA 3: CANTIDAD DE INGRESOS")
print("=".repeat(80))

print("\n📖 SEMÁNTICA:")
print("   Cuenta tickets CREADOS durante [startDate, endDate)")
print("   Es el número de tickets nuevos que entraron al sistema en ese período")

print("\n🔧 IMPLEMENTACIÓN:")

const matchQ3 = {
    createdAt: {
        $gte: startDate,
        $lt: endDate
    }
}

if (estadoFiltro) {
    matchQ3.currentState = estadoFiltro
}

if (clasificadorFiltro) {
    const descendientesQ3 = getDescendantClassifiers(clasificadorFiltro)
    matchQ3["currentClassifications.tipo_solicitud"] = { $in: descendientesQ3 }
}

const totalIngresos = db.tickets.countDocuments(matchQ3)

print(`\n✅ RESULTADO: ${totalIngresos} tickets creados en el período`)

const ejemplosIngresos = db.tickets.find(
    { createdAt: { $gte: startDate, $lt: endDate } }
).sort({ createdAt: 1 }).limit(3).toArray()

print("\n📋 Primeros 3 tickets creados:")
ejemplosIngresos.forEach((ticket, idx) => {
    print(`\n   ${idx + 1}. ID: ${ticket._id}`)
    print(`      Título: ${ticket.title}`)
    print(`      Fecha creación: ${ticket.createdAt.toISOString()}`)
    print(`      Estado actual: ${ticket.currentState}`)
})

print("\n" + "=".repeat(80))
print("CONSULTA 4: CANTIDAD DE CIERRES")
print("=".repeat(80))

print("\n📖 SEMÁNTICA:")
print("   Cuenta TRANSICIONES hacia estado 'closed' durante [startDate, endDate)")
print("   Un ticket puede tener múltiples cierres si fue reabierto")
print("   Solo cuenta cierres que ocurrieron dentro del período especificado")

print("\n🔧 IMPLEMENTACIÓN:")

const pipeline4 = [
    { $unwind: "$history" },
    {
        $match: {
            "history.action": "state_change",
            "history.to": "closed",
            "history.timestamp": {
                $gte: startDate,
                $lt: endDate
            }
        }
    },
    { $count: "totalCierres" }
]

const resultado4 = db.tickets.aggregate(pipeline4).toArray()
const totalCierres = resultado4[0]?.totalCierres || 0

print(`\n✅ RESULTADO: ${totalCierres} cierres en el período`)

const pipeline4detalle = [
    { $unwind: "$history" },
    {
        $match: {
            "history.action": "state_change",
            "history.to": "closed",
            "history.timestamp": {
                $gte: startDate,
                $lt: endDate
            }
        }
    },
    {
        $project: {
            _id: 1,
            title: 1,
            "history.timestamp": 1,
            "history.from": 1,
            "history.comment": 1
        }
    },
    { $limit: 3 }
]

const ejemplosCierres = db.tickets.aggregate(pipeline4detalle).toArray()

print("\n📋 Primeros 3 cierres:")
ejemplosCierres.forEach((item, idx) => {
    print(`\n   ${idx + 1}. Ticket: ${item._id} - ${item.title}`)
    print(`      Fecha cierre: ${item.history.timestamp.toISOString()}`)
    print(`      Estado anterior: ${item.history.from}`)
    if (item.history.comment) {
        print(`      Comentario: ${item.history.comment}`)
    }
})

const ticketsCerrados = db.tickets.countDocuments({
    closedAt: {
        $gte: startDate,
        $lt: endDate
    }
})
print(`\n   📊 Tickets actualmente cerrados que se cerraron en el período: ${ticketsCerrados}`)

print("\n" + "=".repeat(80))
print("CONSULTA 5: LISTA DE ACCIONES")
print("=".repeat(80))

print("\n📖 SEMÁNTICA:")
print("   Retorna todas las acciones del histórico que ocurrieron durante [startDate, endDate)")
print("   Incluye: cambios de estado, comentarios, asignaciones, etc.")
print("   Las acciones pueden ser de cualquier ticket que cumpla los filtros base")

print("\n🔧 IMPLEMENTACIÓN:")

const pipeline5 = [
    {
        $match: {
            "history.timestamp": {
                $gte: startDate,
                $lt: endDate
            }
        }
    },
    { $unwind: "$history" },
    {
        $match: {
            "history.timestamp": {
                $gte: startDate,
                $lt: endDate
            }
        }
    },
    {
        $project: {
            ticketId: "$_id",
            ticketTitle: "$title",
            action: "$history.action",
            timestamp: "$history.timestamp",
            userId: "$history.userId",
            from: "$history.from",
            to: "$history.to",
            comment: "$history.comment",
            assignedTo: "$history.assignedTo"
        }
    },
    { $sort: { timestamp: -1 } },
    { $limit: 50 }
]

const acciones = db.tickets.aggregate(pipeline5).toArray()

print(`\n✅ RESULTADO: ${acciones.length} acciones encontradas (mostrando últimas 50)`)

const accionesPorTipo = {}
acciones.forEach(a => {
    accionesPorTipo[a.action] = (accionesPorTipo[a.action] || 0) + 1
})

print("\n📊 Distribución por tipo de acción:")
Object.entries(accionesPorTipo).forEach(([tipo, count]) => {
    print(`   • ${tipo}: ${count}`)
})

print("\n📋 Últimas 5 acciones:")
acciones.slice(0, 5).forEach((accion, idx) => {
    print(`\n   ${idx + 1}. ${accion.action.toUpperCase()}`)
    print(`      Ticket: ${accion.ticketId} - ${accion.ticketTitle}`)
    print(`      Fecha: ${accion.timestamp.toISOString()}`)
    print(`      Usuario: ${accion.userId}`)

    if (accion.action === 'state_change') {
        print(`      Cambio: ${accion.from} → ${accion.to}`)
    }

    if (accion.action === 'assignment') {
        print(`      Asignado a: ${accion.assignedTo}`)
    }

    if (accion.comment) {
        print(`      Comentario: ${accion.comment}`)
    }
})

print("\n" + "=".repeat(80))
print("📊 RESUMEN EJECUTIVO")
print("=".repeat(80))

const totalTickets = db.tickets.countDocuments({})
const ticketsAbiertos = db.tickets.countDocuments({ currentState: "open" })
const ticketsEnProgreso = db.tickets.countDocuments({ currentState: "in_progress" })
const ticketsActualmenteCerrados = db.tickets.countDocuments({ currentState: "closed" })

print(`\n🎫 TICKETS:`)
print(`   Total en sistema: ${totalTickets}`)
print(`   Abiertos: ${ticketsAbiertos} (${(ticketsAbiertos / totalTickets * 100).toFixed(1)}%)`)
print(`   En progreso: ${ticketsEnProgreso} (${(ticketsEnProgreso / totalTickets * 100).toFixed(1)}%)`)
print(`   Cerrados: ${ticketsActualmenteCerrados} (${(ticketsActualmenteCerrados / totalTickets * 100).toFixed(1)}%)`)

print(`\n📊 MÉTRICAS DEL PERÍODO:`)
print(`   Ingresos: ${totalIngresos} tickets`)
print(`   Cierres: ${totalCierres} transiciones`)
print(`   Reaperturas: ${totalReaperturas} transiciones`)
print(`   Acciones registradas: ${acciones.length}`)

const tasaReapertura = ticketsCerrados > 0 ? (totalReaperturas / ticketsCerrados * 100) : 0
print(`\n📈 INDICADORES:`)
print(`   Tasa de reapertura: ${tasaReapertura.toFixed(1)}%`)
print(`   Tickets con reaperturas: ${ticketsConReaperturas}`)

print("\n" + "=".repeat(80))
print("✅ TODAS LAS CONSULTAS COMPLETADAS EXITOSAMENTE")
print("=".repeat(80) + "\n")

File: streamlit/queries.py
from datetime import datetime
from typing import List, Dict, Any, Optional


class TicketQueries:
    """Clase con todas las queries de MongoDB"""

    def __init__(self, db):
        self.db = db

    def get_descendant_classifiers(self, node_id: str) -> Optional[List[str]]:
        """Obtiene clasificadores descendientes"""
        if node_id == "Todos":
            return None
        descendants = list(self.db.classifiers.find({"ancestors": node_id}))
        return [node_id] + [d["_id"] for d in descendants]

    def get_all_classifiers(self) -> List[str]:
        """Obtiene todos los clasificadores"""
        classifiers = list(
            self.db.classifiers.find({"level": {"$gte": 2}}).sort("name", 1)
        )
        return ["Todos"] + [c["_id"] for c in classifiers]

    def build_match_query(
        self,
        start_date: datetime,
        end_date: datetime,
        estado: str,
        clasificador: str,
        only_date_range: bool = False,
    ) -> Dict[str, Any]:
        """Construye query de match base"""
        query = {}

        if estado != "Todos":
            query["currentState"] = estado

        if clasificador != "Todos":
            descendientes = self.get_descendant_classifiers(clasificador)
            if descendientes:
                query["currentClassifications.tipo_solicitud"] = {"$in": descendientes}

        if only_date_range:
            query["createdAt"] = {"$gte": start_date, "$lte": end_date}

        return query

    def count_tickets_by_state(
        self,
        state: str,
        match_query: Dict[str, Any],
        total_count: int,
        estado_filtro: str,
    ) -> int:
        """Cuenta tickets por estado"""
        if estado_filtro != "Todos":
            return total_count if estado_filtro == state else 0
        return self.db.tickets.count_documents({**match_query, "currentState": state})

    def get_metrics(self, start_date, end_date, estado, clasificador):
        """Obtiene todas las métricas - AHORA CON FILTRO DE FECHA"""
        query = self.build_match_query(start_date, end_date, estado, clasificador)

        # Agregar filtro de fecha a la query base
        query["createdAt"] = {"$gte": start_date, "$lte": end_date}

        total = self.db.tickets.count_documents(query)

        # Para contar por estado, necesitamos manejar el filtro de estado correctamente
        if estado != "Todos":
            abiertos = total if estado == "open" else 0
            en_progreso = total if estado == "in_progress" else 0
            cerrados = total if estado == "closed" else 0
        else:
            abiertos = self.db.tickets.count_documents(
                {**query, "currentState": "open"}
            )
            en_progreso = self.db.tickets.count_documents(
                {**query, "currentState": "in_progress"}
            )
            cerrados = self.db.tickets.count_documents(
                {**query, "currentState": "closed"}
            )

        pipeline = [
            {"$match": query},
            {"$group": {"_id": None, "total": {"$sum": "$reopenCount"}}},
        ]
        reaperturas_result = list(self.db.tickets.aggregate(pipeline))
        reaperturas = reaperturas_result[0]["total"] if reaperturas_result else 0

        return {
            "total": total,
            "abiertos": abiertos,
            "en_progreso": en_progreso,
            "cerrados": cerrados,
            "reaperturas": reaperturas,
        }

    def get_tickets_by_classifier(self, start_date, end_date, estado, clasificador):
        """Obtiene tickets agrupados por clasificador - CON FILTRO DE FECHA"""
        query = self.build_match_query(start_date, end_date, estado, clasificador)

        # Agregar filtro de fecha
        query["createdAt"] = {"$gte": start_date, "$lte": end_date}

        pipeline = [
            {"$match": query},
            {
                "$group": {
                    "_id": "$currentClassifications.tipo_solicitud",
                    "count": {"$sum": 1},
                }
            },
            {"$sort": {"count": -1}},
            {"$limit": 10},
        ]

        return list(self.db.tickets.aggregate(pipeline))

    def get_tickets_trend(self, start_date, end_date, estado, clasificador):
        """Obtiene tendencia de creación de tickets"""
        query = self.build_match_query(
            start_date, end_date, estado, clasificador, only_date_range=True
        )

        pipeline = [
            {"$match": query},
            {
                "$group": {
                    "_id": {
                        "year": {"$year": "$createdAt"},
                        "month": {"$month": "$createdAt"},
                        "day": {"$dayOfMonth": "$createdAt"},
                    },
                    "count": {"$sum": 1},
                }
            },
            {"$sort": {"_id": 1}},
        ]

        return list(self.db.tickets.aggregate(pipeline))

    def get_tickets_list(self, start_date, end_date, estado, clasificador, limit=50):
        """Obtiene lista de tickets"""
        match_query = {
            "createdAt": {"$lt": end_date},
            "$or": [
                {"closedAt": {"$exists": False}},
                {"closedAt": {"$gte": start_date}},
            ],
        }

        if clasificador != "Todos":
            descendientes = self.get_descendant_classifiers(clasificador)
            if descendientes:
                match_query["currentClassifications.tipo_solicitud"] = {
                    "$in": descendientes
                }

        if estado != "Todos":
            match_query["currentState"] = estado

        return list(
            self.db.tickets.find(match_query).sort("createdAt", -1).limit(limit)
        )

    def get_reopenings_stats(self, estado, clasificador):
        """Obtiene estadísticas de reaperturas"""
        query = self.build_match_query(None, None, estado, clasificador)
        total = self.db.tickets.count_documents(query)
        con_reaperturas = self.db.tickets.count_documents(
            {**query, "reopenCount": {"$gt": 0}}
        )
        return {
            "con_reaperturas": con_reaperturas,
            "sin_reaperturas": total - con_reaperturas,
            "total": total,
        }

    def get_resolution_time(self, start_date, end_date, estado, clasificador):
        """Obtiene tiempo de resolución"""
        query = self.build_match_query(start_date, end_date, estado, clasificador)

        pipeline = [
            {
                "$match": {
                    **query,
                    "closedAt": {"$exists": True},
                    "createdAt": {"$gte": start_date, "$lte": end_date},
                }
            },
            {
                "$project": {
                    "tiempo": {
                        "$divide": [
                            {"$subtract": ["$closedAt", "$createdAt"]},
                            1000 * 60 * 60 * 24,
                        ]
                    }
                }
            },
            {
                "$group": {
                    "_id": None,
                    "promedio": {"$avg": "$tiempo"},
                    "minimo": {"$min": "$tiempo"},
                    "maximo": {"$max": "$tiempo"},
                }
            },
        ]

        result = list(self.db.tickets.aggregate(pipeline))
        return result[0] if result else None

    def get_closures_in_period(self, start_date, end_date, clasificador):
        """Obtiene cantidad de cierres en período"""
        query = {}
        if clasificador != "Todos":
            descendientes = self.get_descendant_classifiers(clasificador)
            if descendientes:
                query["currentClassifications.tipo_solicitud"] = {"$in": descendientes}

        pipeline = [
            {"$match": query} if query else {"$match": {}},
            {"$unwind": "$history"},
            {
                "$match": {
                    "history.action": "state_change",
                    "history.to": "closed",
                    "history.timestamp": {"$gte": start_date, "$lte": end_date},
                }
            },
            {"$count": "total"},
        ]

        result = list(self.db.tickets.aggregate(pipeline))
        return result[0]["total"] if result else 0

    def get_reopenings_in_period(self, start_date, end_date, clasificador):
        """Obtiene cantidad de reaperturas en período"""
        query = {}
        if clasificador != "Todos":
            descendientes = self.get_descendant_classifiers(clasificador)
            if descendientes:
                query["currentClassifications.tipo_solicitud"] = {"$in": descendientes}

        pipeline = [
            {"$match": query} if query else {"$match": {}},
            {"$unwind": "$history"},
            {
                "$match": {
                    "history.action": "state_change",
                    "history.from": "closed",
                    "history.to": {"$ne": "closed"},
                    "history.timestamp": {"$gte": start_date, "$lte": end_date},
                }
            },
            {"$count": "total"},
        ]

        result = list(self.db.tickets.aggregate(pipeline))
        return result[0]["total"] if result else 0

    def get_recent_actions(self, start_date, end_date, clasificador, limit=100):
        """Obtiene acciones recientes"""
        query = {"history.timestamp": {"$gte": start_date, "$lte": end_date}}

        if clasificador != "Todos":
            descendientes = self.get_descendant_classifiers(clasificador)
            if descendientes:
                query["currentClassifications.tipo_solicitud"] = {"$in": descendientes}

        pipeline = [
            {"$match": query},
            {"$unwind": "$history"},
            {"$match": {"history.timestamp": {"$gte": start_date, "$lte": end_date}}},
            {
                "$project": {
                    "ticketId": "$_id",
                    "titulo": "$title",
                    "accion": "$history.action",
                    "timestamp": "$history.timestamp",
                    "usuario": "$history.userId",
                    "desde": "$history.from",
                    "hacia": "$history.to",
                }
            },
            {"$sort": {"timestamp": -1}},
            {"$limit": limit},
        ]

        return list(self.db.tickets.aggregate(pipeline))
